File: web/src/utils/readiness.ts
import type { ReadinessArea, ReadinessCheck, ReadinessResponse, ReadinessSeverity } from '@/api/types'

type Translate = (key: string, params?: Record<string, unknown>) => string
type StatusType = 'success' | 'warning' | 'error' | 'info'

export interface ReadinessDescription {
  title: string
  summary: string
  details: string[]
}

export interface ReadinessOverview {
  type: StatusType
  badgeLabel: string
  title: string
  description: string
}

function stringFact(check: ReadinessCheck, key: string): string {
  const value = check.facts?.[key]
  return typeof value === 'string' ? value : ''
}

function numberFact(check: ReadinessCheck, key: string): number {
  const value = check.facts?.[key]
  return typeof value === 'number' ? value : 0
}

function boolFact(check: ReadinessCheck, key: string): boolean {
  return check.facts?.[key] === true
}

function stringArrayFact(check: ReadinessCheck, key: string): string[] {
  const value = check.facts?.[key]
  return Array.isArray(value) ? value.filter((item): item is string => typeof item === 'string') : []
}

function joinItems(items: string[]): string {
  return items.join(', ')
}

export function readinessSeverityType(severity: ReadinessSeverity): Exclude<StatusType, 'info'> {
  switch (severity) {
    case 'blocking':
      return 'error'
    case 'warning':
      return 'warning'
    default:
      return 'success'
  }
}

export function readinessSeverityLabel(t: Translate, severity: ReadinessSeverity): string {
  switch (severity) {
    case 'blocking':
      return t('readiness.severity.blocking')
    case 'warning':
      return t('readiness.severity.warning')
    default:
      return t('readiness.severity.ok')
  }
}

export function readinessAreaLabel(t: Translate, area: ReadinessArea): string {
  return t(`readiness.area.${area}`)
}

export function describeReadinessOverview(t: Translate, readiness: ReadinessResponse | null): ReadinessOverview {
  if (!readiness) {
    return {
      type: 'info',
      badgeLabel: t('readiness.overview.unavailableBadge'),
      title: t('readiness.overview.unavailableTitle'),
      description: t('readiness.overview.unavailableBody')
    }
  }

  if (readiness.blockingCount > 0) {
    return {
      type: 'error',
      badgeLabel: t('readiness.overview.blockingBadge'),
      title: t('readiness.overview.blockingTitle'),
      description: t('readiness.overview.blockingBody', { count: readiness.blockingCount })
    }
  }

  if (readiness.warningCount > 0) {
    return {
      type: 'warning',
      badgeLabel: t('readiness.overview.warningBadge'),
      title: t('readiness.overview.warningTitle'),
      description: t('readiness.overview.warningBody', { count: readiness.warningCount })
    }
  }

  return {
    type: 'success',
    badgeLabel: t('readiness.overview.healthyBadge'),
    title: t('readiness.overview.healthyTitle'),
    description: t('readiness.overview.healthyBody')
  }
}

export function describeReadinessCheck(t: Translate, check: ReadinessCheck): ReadinessDescription {
  switch (check.key) {
    case 'config_path':
      return describeConfigPath(t, check)
    case 'config_sections':
      return describeConfigSections(t, check)
    case 'subscriptions':
      return describeSubscriptions(t, check)
    case 'probing':
      return describeProbing(t, check)
    case 'node_pool':
      return describeNodePool(t, check)
    case 'tun':
      return describeTun(t, check)
    case 'updates':
      return describeUpdates(t, check)
    default:
      return {
        title: check.key,
        summary: t('readiness.checks.unknown.summary'),
        details: []
      }
  }
}

function describeConfigPath(t: Translate, check: ReadinessCheck): ReadinessDescription {
  const path = stringFact(check, 'path')
  const status = stringFact(check, 'status')
  const error = stringFact(check, 'error')
  const details = path ? [t('readiness.checks.configPath.pathDetail', { path })] : []
  if (error) {
    details.push(t('readiness.checks.configPath.errorDetail', { error }))
  }

  let summary = t('readiness.checks.configPath.ok', { path })
  if (status === 'missing') {
    summary = t('readiness.checks.configPath.missing')
  } else if (status === 'not_found') {
    summary = t('readiness.checks.configPath.notFound', { path })
  } else if (status === 'invalid') {
    summary = t('readiness.checks.configPath.invalid', { path })
  }

  return {
    title: t('readiness.checks.configPath.title'),
    summary,
    details
  }
}

function describeConfigSections(t: Translate, check: ReadinessCheck): ReadinessDescription {
  const missingSections = stringArrayFact(check, 'missingSections')
  const missingStatsFlags = stringArrayFact(check, 'missingStatsFlags')
  const missingServices = stringArrayFact(check, 'missingServices')
  const status = stringFact(check, 'status')
  const details: string[] = []

  if (missingSections.length > 0) {
    details.push(t('readiness.checks.configSections.missingSections', { items: joinItems(missingSections) }))
  }
  if (missingStatsFlags.length > 0) {
    details.push(t('readiness.checks.configSections.missingStatsFlags', { items: joinItems(missingStatsFlags) }))
  }
  if (missingServices.length > 0) {
    details.push(t('readiness.checks.configSections.missingServices', { items: joinItems(missingServices) }))
  }

  let summary = t('readiness.checks.configSections.ok')
  if (status === 'unavailable') {
    summary = t('readiness.checks.configSections.unavailable')
  } else if (missingSections.length > 0) {
    summary = t('readiness.checks.configSections.missingSectionsSummary')
  } else if (missingStatsFlags.length > 0 || missingServices.length > 0) {
    summary = t('readiness.checks.configSections.partialSummary')
  }

  return {
    title: t('readiness.checks.configSections.title'),
    summary,
    details
  }
}

function describeSubscriptions(t: Translate, check: ReadinessCheck): ReadinessDescription {
  const status = stringFact(check, 'status')
  const subscriptionCount = numberFact(check, 'subscriptionCount')
  const nodeCount = numberFact(check, 'nodeCount')
  const details = [
    t('readiness.checks.subscriptions.poolDetail', {
      subscriptions: subscriptionCount,
      nodes: nodeCount,
      active: numberFact(check, 'activeCount'),
      staging: numberFact(check, 'stagingCount'),
      candidate: numberFact(check, 'candidateCount'),
      quarantine: numberFact(check, 'quarantineCount'),
      removed: numberFact(check, 'removedCount')
    })
  ]

  let summary = t('readiness.checks.subscriptions.ok', { subscriptions: subscriptionCount, nodes: nodeCount })
  if (status === 'unavailable') {
    summary = t('readiness.checks.subscriptions.unavailable')
  } else if (status === 'empty') {
    summary = t('readiness.checks.subscriptions.empty')
  } else if (status === 'no_nodes') {
    summary = t('readiness.checks.subscriptions.noNodes', { subscriptions: subscriptionCount })
  }

  return {
    title: t('readiness.checks.subscriptions.title'),
    summary,
    details
  }
}

function describeProbing(t: Translate, check: ReadinessCheck): ReadinessDescription {
  const status = stringFact(check, 'status')
  const probeUrl = stringFact(check, 'probeUrl')
  const intervalSec = numberFact(check, 'probeIntervalSec')
  const tagCount = numberFact(check, 'tagCount')
  const details = [t('readiness.checks.probing.configDetail', { probeUrl, intervalSec, tagCount })]

  let summary = t('readiness.checks.probing.running', { tagCount, intervalSec })
  if (status === 'unavailable') {
    summary = t('readiness.checks.probing.unavailable')
  } else if (status === 'not_started') {
    summary = t('readiness.checks.probing.notStarted')
  } else if (status === 'dispatcher_unavailable') {
    summary = t('readiness.checks.probing.dispatcherUnavailable')
  } else if (status === 'idle') {
    summary = t('readiness.checks.probing.idle')
  } else if (status === 'stopped') {
    summary = t('readiness.checks.probing.stopped')
  }

  return {
    title: t('readiness.checks.probing.title'),
    summary,
    details
  }
}

function describeNodePool(t: Translate, check: ReadinessCheck): ReadinessDescription {
  const status = stringFact(check, 'status')
  const activeCount = numberFact(check, 'activeCount')
  const minActiveNodes = numberFact(check, 'minActiveNodes')
  const details = [
    t('readiness.checks.nodePool.distributionDetail', {
      active: activeCount,
      candidate: numberFact(check, 'candidateCount'),
      staging: numberFact(check, 'stagingCount'),
      quarantine: numberFact(check, 'quarantineCount'),
      removed: numberFact(check, 'removedCount')
    })
  ]

  let summary = t('readiness.checks.nodePool.ok', { active: activeCount, minimum: minActiveNodes })
  if (status === 'unavailable') {
    summary = t('readiness.checks.nodePool.unavailable')
  } else if (status === 'empty') {
    summary = t('readiness.checks.nodePool.empty')
  } else if (status === 'below_minimum') {
    summary = t('readiness.checks.nodePool.belowMinimum', { active: activeCount, minimum: minActiveNodes })
  }

  return {
    title: t('readiness.checks.nodePool.title'),
    summary,
    details
  }
}

function describeTun(t: Translate, check: ReadinessCheck): ReadinessDescription {
  const status = stringFact(check, 'status')
  const message = stringFact(check, 'message')
  const details: string[] = []

  if (message) {
    details.push(t('readiness.checks.tun.messageDetail', { message }))
  }
  if (!boolFact(check, 'helperExists')) {
    details.push(t('readiness.checks.tun.helperMissingDetail'))
  }
  if (boolFact(check, 'privilegeInstallRecommended')) {
    details.push(t('readiness.checks.tun.privilegeRecommendedDetail'))
  }
  const machineState = stringFact(check, 'machineState')
  if (machineState) {
    details.push(t('readiness.checks.tun.machineStateDetail', { state: machineState }))
  }

  let summary = boolFact(check, 'running') ? t('readiness.checks.tun.running') : t('readiness.checks.tun.stopped')
  if (status === 'unavailable') {
    summary = t('readiness.checks.tun.unavailable')
  } else if (check.severity === 'blocking') {
    summary = t('readiness.checks.tun.degraded')
  } else if (check.severity === 'warning') {
    summary = t('readiness.checks.tun.warning')
  }

  return {
    title: t('readiness.checks.tun.title'),
    summary,
    details
  }
}

function describeUpdates(t: Translate, check: ReadinessCheck): ReadinessDescription {
  const status = stringFact(check, 'status')
  const currentVersion = stringFact(check, 'currentVersion')
  const latestVersion = stringFact(check, 'latestVersion')
  const source = stringFact(check, 'source')
  const message = stringFact(check, 'message')
  const details: string[] = []

  if (source) {
    details.push(t('readiness.checks.updates.sourceDetail', { source }))
  }
  if (message) {
    details.push(t('readiness.checks.updates.messageDetail', { message }))
  }

  let summary = t('readiness.checks.updates.ok', { currentVersion })
  if (status === 'unavailable') {
    summary = t('readiness.checks.updates.unavailable')
  } else if (status === 'error') {
    summary = t('readiness.checks.updates.error')
  } else if (status === 'stale') {
    summary = t('readiness.checks.updates.stale')
  } else if (boolFact(check, 'updateAvailable')) {
    summary = t('readiness.checks.updates.updateAvailable', { currentVersion, latestVersion })
  }

  return {
    title: t('readiness.checks.updates.title'),
    summary,
    details
  }
}

File: app/webpanel/readiness.go
package webpanel

import (
	"context"
	"encoding/json"
	"os"
	"strings"
	"time"
)

type ReadinessSeverity string

const (
	ReadinessSeverityOK       ReadinessSeverity = "ok"
	ReadinessSeverityWarning  ReadinessSeverity = "warning"
	ReadinessSeverityBlocking ReadinessSeverity = "blocking"
)

type ReadinessArea string

const (
	ReadinessAreaConfig        ReadinessArea = "config"
	ReadinessAreaSubscriptions ReadinessArea = "subscriptions"
	ReadinessAreaNodePool      ReadinessArea = "node_pool"
	ReadinessAreaTun           ReadinessArea = "tun"
	ReadinessAreaRuntime       ReadinessArea = "runtime"
	ReadinessAreaUpdates       ReadinessArea = "updates"
)

type ReadinessCheck struct {
	Key         string                 `json:"key"`
	Area        ReadinessArea          `json:"area"`
	Severity    ReadinessSeverity      `json:"severity"`
	ActionRoute string                 `json:"actionRoute,omitempty"`
	Facts       map[string]interface{} `json:"facts,omitempty"`
}

type ReadinessResponse struct {
	Healthy       bool             `json:"healthy"`
	BlockingCount int              `json:"blockingCount"`
	WarningCount  int              `json:"warningCount"`
	UpdatedAt     string           `json:"updatedAt"`
	Checks        []ReadinessCheck `json:"checks"`
}

func (wp *WebPanel) readinessSnapshot(ctx context.Context) ReadinessResponse {
	checks := make([]ReadinessCheck, 0, 7)
	configMap, configPathStatus, configPathFacts := wp.readinessLoadConfig()
	checks = append(checks, ReadinessCheck{
		Key:         "config_path",
		Area:        ReadinessAreaConfig,
		Severity:    configPathStatus,
		ActionRoute: "/config",
		Facts:       configPathFacts,
	})
	checks = append(checks, wp.readinessConfigSectionsCheck(configMap))
	checks = append(checks, wp.readinessSubscriptionsCheck())
	checks = append(checks, wp.readinessProbingCheck())
	checks = append(checks, wp.readinessNodePoolCheck())
	checks = append(checks, wp.readinessTunCheck())
	checks = append(checks, wp.readinessUpdatesCheck(ctx))

	response := ReadinessResponse{
		Healthy:   true,
		UpdatedAt: time.Now().UTC().Format(time.RFC3339),
		Checks:    checks,
	}
	for _, check := range checks {
		switch check.Severity {
		case ReadinessSeverityBlocking:
			response.BlockingCount++
			response.Healthy = false
		case ReadinessSeverityWarning:
			response.WarningCount++
		}
	}
	return response
}

func (wp *WebPanel) readinessLoadConfig() (map[string]interface{}, ReadinessSeverity, map[string]interface{}) {
	configPath := strings.TrimSpace(wp.config.ConfigPath)
	facts := map[string]interface{}{
		"path": configPath,
	}
	if configPath == "" {
		facts["status"] = "missing"
		return nil, ReadinessSeverityBlocking, facts
	}

	info, err := os.Stat(configPath)
	if err != nil {
		facts["status"] = "not_found"
		facts["error"] = err.Error()
		return nil, ReadinessSeverityBlocking, facts
	}
	facts["sizeBytes"] = info.Size()

	cfm := NewConfigFileManager(configPath)
	raw, err := cfm.ReadConfig()
	if err != nil {
		facts["status"] = "invalid"
		facts["error"] = err.Error()
		return nil, ReadinessSeverityBlocking, facts
	}

	var configMap map[string]interface{}
	if err := json.Unmarshal(raw, &configMap); err != nil {
		facts["status"] = "invalid"
		facts["error"] = err.Error()
		return nil, ReadinessSeverityBlocking, facts
	}

	facts["status"] = "ok"
	return configMap, ReadinessSeverityOK, facts
}

func (wp *WebPanel) readinessConfigSectionsCheck(config map[string]interface{}) ReadinessCheck {
	facts := map[string]interface{}{}
	if config == nil {
		facts["status"] = "unavailable"
		return ReadinessCheck{
			Key:         "config_sections",
			Area:        ReadinessAreaConfig,
			Severity:    ReadinessSeverityBlocking,
			ActionRoute: "/config",
			Facts:       facts,
		}
	}

	missingSections := make([]string, 0, 3)
	missingStatsFlags := make([]string, 0, 4)

	if _, ok := config["api"].(map[string]interface{}); !ok {
		missingSections = append(missingSections, "api")
	}
	if _, ok := config["stats"].(map[string]interface{}); !ok {
		missingSections = append(missingSections, "stats")
	}

	policy, _ := config["policy"].(map[string]interface{})
	systemPolicy, _ := policy["system"].(map[string]interface{})
	for _, key := range []string{
		"statsInboundUplink",
		"statsInboundDownlink",
		"statsOutboundUplink",
		"statsOutboundDownlink",
	} {
		value, ok := systemPolicy[key].(bool)
		if !ok || !value {
			missingStatsFlags = append(missingStatsFlags, key)
		}
	}

	services := make([]string, 0)
	apiSection, _ := config["api"].(map[string]interface{})
	if rawServices, ok := apiSection["services"].([]interface{}); ok {
		for _, service := range rawServices {
			if value, ok := service.(string); ok && strings.TrimSpace(value) != "" {
				services = append(services, strings.TrimSpace(value))
			}
		}
	}
	missingServices := missingRequiredServices(services)

	facts["status"] = "ok"
	if len(missingSections) > 0 {
		facts["status"] = "missing_sections"
		facts["missingSections"] = missingSections
	}
	if len(missingStatsFlags) > 0 {
		facts["missingStatsFlags"] = missingStatsFlags
		if facts["status"] == "ok" {
			facts["status"] = "missing_stats_flags"
		}
	}
	if len(missingServices) > 0 {
		facts["missingServices"] = missingServices
		if facts["status"] == "ok" {
			facts["status"] = "missing_services"
		}
	}

	severity := ReadinessSeverityOK
	if len(missingSections) > 0 {
		severity = ReadinessSeverityBlocking
	} else if len(missingStatsFlags) > 0 || len(missingServices) > 0 {
		severity = ReadinessSeverityWarning
	}

	return ReadinessCheck{
		Key:         "config_sections",
		Area:        ReadinessAreaConfig,
		Severity:    severity,
		ActionRoute: "/config",
		Facts:       facts,
	}
}

func missingRequiredServices(services []string) []string {
	required := []string{
		"HandlerService",
		"StatsService",
		"LoggerService",
		"RoutingService",
		"ObservatoryService",
	}
	if len(services) == 0 {
		return append([]string{}, required...)
	}

	existing := make(map[string]struct{}, len(services))
	for _, service := range services {
		existing[strings.TrimSpace(service)] = struct{}{}
	}

	missing := make([]string, 0, len(required))
	for _, service := range required {
		if _, ok := existing[service]; !ok {
			missing = append(missing, service)
		}
	}
	return missing
}

func (wp *WebPanel) readinessSubscriptionsCheck() ReadinessCheck {
	if wp.subManager == nil {
		return ReadinessCheck{
			Key:         "subscriptions",
			Area:        ReadinessAreaSubscriptions,
			Severity:    ReadinessSeverityBlocking,
			ActionRoute: "/subscriptions",
			Facts: map[string]interface{}{
				"status": "unavailable",
			},
		}
	}

	snapshot := wp.subManager.ReadinessSnapshot()
	facts := map[string]interface{}{
		"subscriptionCount": snapshot.SubscriptionCount,
		"nodeCount":         snapshot.NodeCount,
		"activeCount":       snapshot.PoolSummary.ActiveCount,
		"candidateCount":    snapshot.PoolSummary.CandidateCount,
		"stagingCount":      snapshot.PoolSummary.StagingCount,
		"quarantineCount":   snapshot.PoolSummary.QuarantineCount,
		"removedCount":      snapshot.PoolSummary.RemovedCount,
	}

	severity := ReadinessSeverityOK
	status := "ok"
	switch {
	case snapshot.SubscriptionCount == 0:
		severity = ReadinessSeverityWarning
		status = "empty"
	case snapshot.NodeCount == 0:
		severity = ReadinessSeverityWarning
		status = "no_nodes"
	}
	facts["status"] = status

	return ReadinessCheck{
		Key:         "subscriptions",
		Area:        ReadinessAreaSubscriptions,
		Severity:    severity,
		ActionRoute: "/subscriptions",
		Facts:       facts,
	}
}

func (wp *WebPanel) readinessProbingCheck() ReadinessCheck {
	if wp.subManager == nil {
		return ReadinessCheck{
			Key:         "probing",
			Area:        ReadinessAreaRuntime,
			Severity:    ReadinessSeverityBlocking,
			ActionRoute: "/node-pool",
			Facts: map[string]interface{}{
				"status": "unavailable",
			},
		}
	}

	snapshot := wp.subManager.ReadinessSnapshot()
	facts := map[string]interface{}{
		"started":             snapshot.Started,
		"dispatcherAvailable": snapshot.DispatcherAvailable,
		"probeUrl":            snapshot.ProbeURL,
		"probeIntervalSec":    snapshot.ProbeIntervalSec,
		"tagCount":            snapshot.Prober.TagCount,
		"running":             snapshot.Prober.Running,
	}

	severity := ReadinessSeverityOK
	status := "ok"
	switch {
	case !snapshot.Started:
		severity = ReadinessSeverityWarning
		status = "not_started"
	case !snapshot.DispatcherAvailable:
		severity = ReadinessSeverityWarning
		status = "dispatcher_unavailable"
	case snapshot.Prober.TagCount == 0:
		severity = ReadinessSeverityWarning
		status = "idle"
	case !snapshot.Prober.Running:
		severity = ReadinessSeverityWarning
		status = "stopped"
	default:
		status = "running"
	}
	facts["status"] = status

	return ReadinessCheck{
		Key:         "probing",
		Area:        ReadinessAreaRuntime,
		Severity:    severity,
		ActionRoute: "/node-pool",
		Facts:       facts,
	}
}

func (wp *WebPanel) readinessNodePoolCheck() ReadinessCheck {
	if wp.subManager == nil {
		return ReadinessCheck{
			Key:         "node_pool",
			Area:        ReadinessAreaNodePool,
			Severity:    ReadinessSeverityBlocking,
			ActionRoute: "/node-pool",
			Facts: map[string]interface{}{
				"status": "unavailable",
			},
		}
	}

	snapshot := wp.subManager.ReadinessSnapshot()
	summary := snapshot.PoolSummary
	facts := map[string]interface{}{
		"status":          "ok",
		"healthy":         summary.Healthy,
		"activeCount":     summary.ActiveCount,
		"candidateCount":  summary.CandidateCount,
		"stagingCount":    summary.StagingCount,
		"quarantineCount": summary.QuarantineCount,
		"removedCount":    summary.RemovedCount,
		"minActiveNodes":  summary.MinActiveNodes,
	}

	severity := ReadinessSeverityOK
	switch {
	case snapshot.NodeCount == 0:
		severity = ReadinessSeverityWarning
		facts["status"] = "empty"
	case !summary.Healthy:
		severity = ReadinessSeverityWarning
		facts["status"] = "below_minimum"
	}

	return ReadinessCheck{
		Key:         "node_pool",
		Area:        ReadinessAreaNodePool,
		Severity:    severity,
		ActionRoute: "/node-pool",
		Facts:       facts,
	}
}

func (wp *WebPanel) readinessTunCheck() ReadinessCheck {
	if wp.tunManager == nil {
		return ReadinessCheck{
			Key:         "tun",
			Area:        ReadinessAreaTun,
			Severity:    ReadinessSeverityWarning,
			ActionRoute: "/settings",
			Facts: map[string]interface{}{
				"status": "unavailable",
			},
		}
	}

	status := wp.tunStatusSnapshotWithoutEgressProbe()
	facts := map[string]interface{}{
		"status":                      status.Status,
		"running":                     status.Running,
		"available":                   status.Available,
		"message":                     status.Message,
		"helperExists":                status.HelperExists,
		"elevationReady":              status.ElevationReady,
		"privilegeInstallRecommended": status.PrivilegeInstallRecommended,
		"machineState":                status.MachineState,
	}

	severity := ReadinessSeverityOK
	switch {
	case status.MachineState == MachineStateDegraded || status.Status == "error":
		severity = ReadinessSeverityBlocking
	case status.Status == "blocked" || status.Status == "unavailable" || status.PrivilegeInstallRecommended || !status.HelperExists:
		severity = ReadinessSeverityWarning
	}

	return ReadinessCheck{
		Key:         "tun",
		Area:        ReadinessAreaTun,
		Severity:    severity,
		ActionRoute: "/settings",
		Facts:       facts,
	}
}

func (wp *WebPanel) readinessUpdatesCheck(ctx context.Context) ReadinessCheck {
	if wp.releaseChecker == nil {
		return ReadinessCheck{
			Key:         "updates",
			Area:        ReadinessAreaUpdates,
			Severity:    ReadinessSeverityWarning,
			ActionRoute: "/dashboard",
			Facts: map[string]interface{}{
				"status": "unavailable",
			},
		}
	}

	checkCtx := ctx
	if checkCtx == nil {
		checkCtx = context.Background()
	}
	checkCtx, cancel := context.WithTimeout(checkCtx, 2*time.Second)
	defer cancel()

	status := wp.releaseChecker.Check(checkCtx, false)
	facts := map[string]interface{}{
		"status":            status.Status,
		"source":            status.Source,
		"currentVersion":    status.CurrentVersion,
		"latestVersion":     status.LatestVersion,
		"updateAvailable":   status.UpdateAvailable,
		"message":           status.Message,
		"latestPublishedAt": status.LatestPublishedAt,
	}

	severity := ReadinessSeverityOK
	if status.Status == "error" || status.Status == "stale" || status.UpdateAvailable {
		severity = ReadinessSeverityWarning
	}

	return ReadinessCheck{
		Key:         "updates",
		Area:        ReadinessAreaUpdates,
		Severity:    severity,
		ActionRoute: "/dashboard",
		Facts:       facts,
	}
}

func readinessCheckByKey(checks []ReadinessCheck, key string) (ReadinessCheck, bool) {
	for _, check := range checks {
		if check.Key == key {
			return check, true
		}
	}
	return ReadinessCheck{}, false
}
